chore(deps): update reqwest

flavio opened the pull request:
_Issue #, if available:_

The current version of tough depends on an older version of reqwest which in cause a old version of hyper to be consumed.

This hyper upgrade is massive, having a unified version of hyper inside of projects consuming the tough crate is beneficial. For example, hyper testing frameworks can work only with one version of the library at a time.

_Description of changes:_

Update to latest version of reqwest, v0.12. This version depends on hyper v1, while previous versions (like v0.11) rely on hyper v0.14.

flavio commented:
I'll look into the failed unit tests, sorry... I didn't spot them locally

webern commented:
> I'll look into the failed unit tests, sorry... I didn't spot them locally

It's an unfortunate lint that we have using Cargo Deny. The lint causes an error when two major versions of a dependency exist in Cargo.lock. To fix the lint, the list of ignored dependencies has to be updated in `deny.toml` in the `bans` `skip` section https://github.com/awslabs/tough/blob/eb5e25e1609ac30d6c9c5f694ccc608acc35401c/deny.toml#L69

`make check-licenses` will run this locally https://github.com/awslabs/tough/blob/eb5e25e1609ac30d6c9c5f694ccc608acc35401c/Makefile#L16

flavio commented:
@webern I've updated `cargo.deny`. Unfortunately the list of duplicated crates is quite bit. Quite some of them come from crates related with `aws-smithy`. Maybe you have influence over them?

flavio commented:
@webern sorry, this should be fixed now

webern commented:
> sorry, this should be fixed now

Can you also turn off auto-formatting of TOML files and restore them so that the PR has the minimum possible diff? We prefer to avoid auto-formatting churn in our git history.

Thank you!

flavio committed:
chore(deps): update reqwest

Update to latest version of reqwest, v0.12. This version depends on
hyper v1, while previous versions (like v0.11) rely on hyper v0.14.

This hyper upgrade is massive, having a unified version of hyper inside
of projects consuming the tough crate is beneficial. For example, hyper
testing frameworks can work only with one version of the library at a
time.

`cargo deny` had to be configured to ignore some duplicated crates.
The reqwest update is massive, is the pinnacle of a major update of hyper.
A lot of hyper dependencies received major bumps too.

Some dependencies of tough (including development only ones), have not
been updated to consume these new crates.

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
@webern sorry about that, I fixed it.

Moreover, `make ci` completes successfully on my computer now.

webern commented:
Looks good, thank you for dealing with our `deny.toml` :roll_eyes:

flavio commented:
Sorry to bother you, do you have any ETA about when you plan to merge this PR and even tag a new release of the crate? :pray:

jpculp commented:
Sorry for the radio silence here. I was concerned about the number of duplicated crates that were being pulled in. I'm going to poke at this early next week and see how things look once we bump smithy and aws-sdk-rust. Do you mind if I add your commit to that larger dependency update PR, or would you rather rebase this PR on top of it?

flavio commented:
Thanks for looking into that! I'm fine with you adding this commit to a larger dependency PR

jpculp commented:
After bumping some things in #781, I still couldn't seem to get out of the situation where we pull two versions of `hyper`. From what I can tell, support for **v1** is only available in [aws-smithy-experimental](https://crates.io/crates/aws-smithy-experimental). If we wanted to stick with the stable release then anyone using `tough` and `aws-smithy` would end up having the two versions of `hyper` you're looking to avoid. Just so I can understand better, can you elaborate on your use-case for `hyper v1`?

flavio commented:
I need tough to consume latest version of `reqwest` to address [this issue](https://github.com/flavio/tough-hyper-bug).

jpculp commented on tough/Cargo.toml:
```diff
@@ -22,7 +22,7 @@ log = "0.4"
 pem = "3"
 percent-encoding = "2"
-reqwest = { version = "0.11", optional = true, default-features = false, features = ["stream"] }
+reqwest = { version = "0.12", optional = true, default-features = false, features = ["stream"] }
```
```suggestion
reqwest = { version = "0.11", optional = true, default-features = false, features = ["stream"] }
reqwest-hyper-v1 = { package = "reqwest", version = "0.12", optional = true, default-features = false, features = ["stream"] }
```
I wonder if we could do this and separate them by feature. Crates stuck on the old hyper would continue to use `http` and those on the new could use a `http-hyper-v1`. Curious if @bcressey or @cbgbt have thoughts around this as well.

bcressey commented:
> From what I can tell, support for v1 is only available in [aws-smithy-experimental](https://crates.io/crates/aws-smithy-experimental).

From the crate description:
> This crate allows customers to use Hyper 1.0. A valuable consequence of this is access to aws-lc-rs and its FIPS compliant crypto. This is available behind the crypto-aws-lc-fips feature. Note: FIPS support has somewhat [complex build requirements](https://github.com/aws/aws-lc/blob/main/BUILDING.md), namely CMake and Go

This will be required soon for FIPS support in `updog`, so I would like to start moving in that direction.

flavio commented:
Hello, is there any update on bumping the hyper/reqwest?

This is blocking the release of `sigstore-rs` crate :cry:

jpculp commented:
@flavio, sorry about the delay on this. I'm still trying to get some answers on `aws-smithy-experimental`. If it turns out that it is still too early to take advantage of it, what are your thoughts on the feature flag workaround above?

flavio committed:
Introduce feature flag

Signed-off-by: Flavio Castelli <fcastelli@suse.com>

flavio commented:
@jpculp I think using the feature flag is the only way to go, and it works (I tested it).

I've force pushed the following changes:
- Rebase against `development` branch
- Initial patch
- Introduce feature flag

Once you're happy about the changes I'll squash the 2 commits into a single one. I just wanted to make easy to understand the amount of changes requested by the feature flag.

flavio commented:
Fantastic news folks, because of a [happy coincidence](https://github.com/sigstore/sigstore-rs/pull/389) I finally understood why sigstore `main` branch wasn't working with tough anymore.
I wouldn't have been able to understand what was going on without 1) the automated PR triggering my curiosity 2) the proposal made here to include different versions of reqwest

## About this PR

The Sigstore project doesn't need any bump of reqwest by tough. We can solve the issue all on our own (more details below).

Having tough bump reqwest would still be nice, but [this commit](https://github.com/awslabs/tough/pull/769/commits/c903eaad71ef6b5d604e66ebf897fbccd2a5506b) is definitely not needed.

There might still be value inside of the 1st commit, because of the updates to `cargo.deny`. However I think you probably won't need them once the whole "AWS ecosystem" moves to the latest version of reqwest.

Feel free to close this PR.

## The root cause and the fix

It can be found [here](https://github.com/sigstore/sigstore-rs/pull/389/commits/d80b800b17fc5cda5aed84a8435aa7e8aac41add)

jpculp commented:
@flavio, that's fantastic news! On our side we're actually working on the code to jump to `aws-smithy-experimental` for full `hyper` v1 support. After talking to some folks on our end, we found it might not be as heavy of a lift as we were expecting. Thank you for your patience in all of this!

jpculp commented:
It's still in progress, but we're working on the `aws-smithy-experimental` switch in #826.

jpculp commented:
Good news. We merged #826 and should be ready for `reqwest` **0.12** now. Are you able to undo the feature toggle and rebase, or would you rather we just take care of it from here?

ginglis13 commented:
Updated `reqwest` to 0.12 in #828 - thanks for paving the way @flavio !

flavio commented:
@jpculp, @ginglis13 thanks for having handled that. I'm sorry, but the last days have been pretty busy for me
